### PR TITLE
Spring Cloud Config reference document section

### PR DESCRIPTION
**Pull request by joaoandremartins**
Adds the missing Spring Cloud Config section to the Spring Cloud GCP
reference document.

### CODE REVIEW

**Comment by joaoandremartins** [comment]
@jabubake FYI

**Comment by joaoandremartins** [comment]
All -- this is likely the last missing piece for M1, a timely review is appreciated so M1 can be launched asap :)

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 1]
```diff
@@ -0,0 +1,101 @@
+== Spring Cloud Config
+
+Spring Cloud GCP makes it possible to use the
+https://cloud.google.com/deployment-manager/runtime-configurator/reference/rest/[Google
+Runtime Configuration API] as a
+https://cloud.spring.io/spring-cloud-config/[Spring Cloud Config] server to remotely store your
+application configuration data.
+
+The Spring Cloud GCP Config support is provided via a Spring Boot starter that makes notation like
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Config support => Config client support

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+is hosted, if different from the one in the <<spring-cloud-gcp-core,Spring Cloud GCP Core Module>> |
+| `spring.cloud.gcp.config.credentials.location` | OAuth2 credentials for authenticating with the
+Google Runtime Configuration API, if different from the ones in the
+<<spring-cloud-gcp-core,Spring Cloud GCP Core Module>> |
+| `spring.cloud.gcp.config.credentials.scopes` |
+https://developers.google.com/identity/protocols/googlescopes[OAuth2 scope] for Spring Cloud GCP
+Config credentials | https://www.googleapis.com/auth/cloudruntimeconfig
+|===
+
+NOTE: These properties should be specified in a `bootstrap.yml` file, rather than the usual
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: should be we state `boostrap.yml` / `bootstrap.properties`.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+`applications.yml`
+
+=== Quick start
+
+1. Create a configuration in the Google Runtime Configuration API that is called
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Thank you for  clearly explaining this :)

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+`spring.cloud.gcp.config.name`_`spring.cloud.gcp.config.profile`. In other words, if
+`spring.cloud.gcp.config.name` is `myapp` and `spring.cloud.gcp.config.profile` is prod, the
+configuration should be called `myapp_prod`.
++
+In order to do that, you should have the
+https://cloud.google.com/sdk/[Google Cloud SDK] installed, own a Google Cloud Project and run the
+following command:
++
+----
+gcloud beta runtime-config configs create myapp_prod
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: `gcloud init` as well ?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+@Value("${queue_size}")
+private int queueSize;
+----
+
+When your Spring application starts, if an instance of the class that contains the `queueSize`
+field is auto-wired, the `queueSize` field value will be set to 25.
+
+=== Refreshing the configuration at runtime
+
+Using http://cloud.spring.io/spring-cloud-static/docs/1.0.x/spring-cloud.html#_endpoints[Spring
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: remove "Using"

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+Boot Actuator] enables a `/refresh` endpoint in your application that can be used to refresh the
+values of all the Spring Config-managed variables.
+
+To achieve that, add a `@RefreshScope` annotation to your `@Configuration` class(es).
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: s/achieve that/achieve this/

Comment (jabubake) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: I think @meltsufin noted that it does n't have to be a `@Configuration` annotation and can be any auto-wired bean ?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+Then, if you change the value of `myapp_prod`'s `queue_size` variable in the Google Runtime
+Configuration API and then hit the `/refresh` endpoint of your application, you can verify that the
+value of the `queueSize` field is now the changed value.
```
Comment (jabubake) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: s/is now the changed value/has been updated/

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 1 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: I omitted the "client" bit on purpose because I don't want readers to think they can use the generic Spring Cloud Config client to connect to the Google Runtime Config API. The Spring Cloud Config client actually can't be used to get config information from us (which is why it was a good idea in the end to not re-use the Spring Cloud Config property namespace `spring.cloud.config.*`).

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 5 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: I'm afraid that omitting "Using" will cause it to sound like Actuator is enabled by default?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 6 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Ah, touché. I don't know where I read that it had to be a `@Configuration`...

Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Actually, if it isn't an `@Configuration`, the class has to be annotated with something, otherwise it doesn't get picked up... The thing is there might be annotations other than `@Configuration` that also make it possible to auto-wire the class (and not include the `@Configuration` annotation themselves).
I'll reword things here a bit, even though above I say that `@Configuration` should be added, and that should probably stay that way.

Comment (artembilan) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Some answers should be here: http://cloud.spring.io/spring-cloud-static/Dalston.SR4/multi/multi__spring_cloud_context_application_context_services.html#_refresh_scope

> A Spring `@Bean` that is marked as `@RefreshScope` will get special treatment when there is a configuration change. 

And there is definitely an IMPORTANT not to be cautious with the `@Configuration` classes

Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: In my toy app, I'm using a class such as the following one to auto-wire later:

```
@Configuration
public class SampleConfig {

  @Value("${url}")
  private String url;

  public String getUrl() {
    return url;
  }

  public void setUrl(String url) {
    this.url = url;
  }
}
```

Can you recommend an annotation to use other than `@Configuration` that makes beans of that kind auto-wireable?

Comment (markfisher) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: the most generic option is `@Component`

Comment (artembilan) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: ... or nothing on the class and in some `@Configuration` have it like:
```
@Bean
@RefreshScope
SampleConfig myConfig() {
   return new SampleConfig();
}
```

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+[source,subs="normal"]
+----
+dependencies {
+    compile group: 'org.springframework.cloud', name: 'spring-cloud-gcp-starter-config', version: '{project-version}'
+}
+----
+
+=== Configuration
+
+The following parameters are configurable in Spring Cloud GCP Config:
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Which ones are optional and which required? Also, maybe state up here that they should go into `boostrap.properties` and link to what that is.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+Config credentials | https://www.googleapis.com/auth/cloudruntimeconfig
+|===
+
+NOTE: These properties should be specified in a `bootstrap.yml`/`bootstrap.properties` file, rather
+than the usual `applications.yml`/`application.properties`.
+
+=== Quick start
+
+1. Create a configuration in the Google Runtime Configuration API that is called
+`spring.cloud.gcp.config.name`_`spring.cloud.gcp.config.profile`. In other words, if
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: This is a bit confusing. Maybe `${spring.cloud.gcp.config.name}_${spring.coud.gcp.config.profile}` for better clarity?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 1 above)
Comment (jabubake) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: I think in that case, it might best to rephrase the server part as "Configuration server" and not imply that it may be Git backed.

Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Because we don't actually use Spring Cloud Config for anything (i.e., we basically re-implement it), I wonder if it's worth rewording to say something like "storing configuration data in a similar way to Spring Cloud Config"?
I'm also not comfortable with the Spring Cloud Config server bit.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 8 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Added an Optional column to the table and added a hyperlink to `bootstrap.properties`.

**Comment by joaoandremartins** [comment]
I think I resolved every comment, but let me know if there is anything I missed.
Thanks!

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+----
+
+=== Configuration
+
+The following parameters are configurable in Spring Cloud GCP Config:
+
+|===
+| Name | Description | Optional | Default value
+| `spring.cloud.gcp.config.enabled` | Enables the Config client | Yes | `true`
+| `spring.cloud.gcp.config.name` |
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Is it config name or application name? It's confusing because you're linking to docs about `spring.application.name`. Should we just use `spring.application.name` or at least default to it?

Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: They're not related. It's suggested that this is set to `spring.application.name`, however, that's merely a suggestion, users can name the config anything they want.

Comment (meltsufin) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: I think we can omit suggesting any connection between this and `spring.application.name` and also not link to http://cloud.spring.io/spring-cloud-static/spring-cloud-config/1.3.3.RELEASE/single/spring-cloud-config.html#_locating_remote_configuration_resources.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+Name of your application | No |
+| `spring.cloud.gcp.config.profile` |
+Configuration's Spring profile (e.g., prod) | Yes | default
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: default -> `default`

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+| `spring.cloud.gcp.config.timeout` | Timeout in milliseconds for connecting to the Google Runtime
+Configuration API | Yes | 60000
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: 60000 -> `60000`

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 13]
```diff
@@ -0,0 +1,115 @@
+== Spring Cloud Config
+
+Spring Cloud GCP makes it possible to use the
+https://cloud.google.com/deployment-manager/runtime-configurator/reference/rest/[Google
+Runtime Configuration API] as a
+https://cloud.spring.io/spring-cloud-config/[Spring Cloud Config] server to remotely store your
+application configuration data.
+
+The Spring Cloud GCP Config support is provided via a Spring Boot starter that makes notation like
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: I'm not a fan of using the word "notation" here.
Here's how I would rephrase:

The Spring Cloud GCP Config support is provided via a Spring Boot starter that enables the use of the `@Value` annotation for injecting values stored in the Google Runtime Configuration API into fields of Spring-managed beans.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+| Yes |
+| `spring.cloud.gcp.config.credentials.location` | OAuth2 credentials for authenticating with the
+Google Runtime Configuration API, if different from the ones in the
+<<spring-cloud-gcp-core,Spring Cloud GCP Core Module>> | Yes |
+| `spring.cloud.gcp.config.credentials.scopes` |
+https://developers.google.com/identity/protocols/googlescopes[OAuth2 scope] for Spring Cloud GCP
+Config credentials | Yes | https://www.googleapis.com/auth/cloudruntimeconfig
+|===
+
+NOTE: These properties should be specified in a
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Link to some explanation of why or at least what bootstrap configuration is?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+file, rather than the usual `applications.yml`/`application.properties`.
+
+=== Quick start
+
+1. Create a configuration in the Google Runtime Configuration API that is called
+`${spring.cloud.gcp.config.name}_${spring.cloud.gcp.config.profile}`.
+In other words, if `spring.cloud.gcp.config.name` is `myapp` and `spring.cloud.gcp.config.profile`
+is prod, the configuration should be called `myapp_prod`.
++
+In order to do that, you should have the
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: In order to do *what*?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+gcloud beta runtime-config configs variables set queue_size 25 --config-name myapp_prod
+----
+
+2. Configure your bootstrap.yml file with your application's configuration data:
++
+----
+spring.cloud.gcp.config.name=myapp
+spring.cloud.gcp.config.profile=prod
+----
+3. Add a field annotated with `@Value` to a POJO:
```
Comment (meltsufin) [inline spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Why POJO? I think it works with any Spring-managed bean.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 14 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: There is already a link that explains bootstrap configuration pretty well IMO.
Do you think more explanation is due here?

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 15 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: To create the configuration.

File: spring-cloud-gcp-docs/src/main/asciidoc/config.adoc (hunk 14 above)
Comment (meltsufin) [reply spring-cloud-gcp-docs/src/main/asciidoc/config.adoc]: Sorry, I think I just missed it.
